src/monitor_trigger.py

```diff
@@ -0,0 +1,275 @@
+    ]
+    # select from second element of list_dim since can loop over 3D
+    # loop over all combs of dims
+    for i in itertools.product(*[ds[d].values for d in list_dim[1:]]):
+        ds_sel = ds.sel({d: i[k] for k, d in enumerate(list_dim[1:])})
+        ds_sel_upsample = ds_sel.rio.reproject(
+            ds_sel.rio.crs,
+            resolution=resolution,
+            resampling=Resampling.nearest,
+            nodata=np.nan,
+        )
+        upsample_list.append(
+            ds_sel_upsample.expand_dims(
+                {d: [i[k]] for k, d in enumerate(list_dim[1:])}
+            )
+        )
+    ds_upsample = xr.combine_by_coords(upsample_list)
+    # reproject changes spatial dims names to x and y
+    # so change back here
+    ds_upsample = ds_upsample.rename({"x": ds.aat.x_dim, "y": ds.aat.y_dim})
+    return ds_upsample
+
+
+def compute_stats_iri(
+    da: xr.DataArray,
+    gdf_aoi: GeoDataFrame,
+    adm0_col: str,
+    pcode0_col: str,
+    perc_area: float,
+    threshold_bavg: float,
+    threshold_diff: float,
+) -> pd.DataFrame:
+    """
+    Compute statistics on the IRI tercile probability forecast for the AOI.
+
+    Parameters
+    ----------
+    da: xr.DataArray
+        DataArray containing the IRI forecast
+    gdf_aoi: gpd.GeoDataFrame
+        GeoDataFrame of which the outer bounds are the area of interest
+    # TODO: figure out reqs dissolve?
+    adm0_col
+    pcode0_col
+    threshold_bavg
+        minimum probability of below average precipitation for the raster
+        cell to be reaching the threshold
+    threshold_diff
+        minimum difference of the below average and above average
+        probability for the raster cell to be reaching the threshold
+    perc_area
+        minimum percentage of the area within `gdf_aoi` to reach the
+        threshold(s) for the trigger to be met
+
+    Returns
+    -------
+    DataFrame containing the statistics on the AOI.
+    """
+    # dissolve the region to one polygon
+    # i.e. we compute the statistics over the whole region
+    gdf_aoi_dissolved = gdf_aoi.dissolve(by=adm0_col)
+    gdf_aoi_dissolved = gdf_aoi_dissolved[[pcode0_col, "geometry"]]
+
+    # sel below average tercile
+    da_bavg = da.sel(C=0)
+
+    # compute the stats on all data
+    df_stats_aoi_bavg = da_bavg.aat.compute_raster_stats(
+        gdf=gdf_aoi_dissolved,
+        feature_col=pcode0_col,
+        percentile_list=[100 - perc_area],
+    )
+
+    # compute the stats on data above threshold
+    da_bavg_thresh = da_bavg.where(da_bavg >= threshold_bavg)
+    df_stats_aoi_bavg_thresh = da_bavg_thresh.aat.compute_raster_stats(
+        gdf=gdf_aoi_dissolved, feature_col=pcode0_col
+    )
+    # compute perc of region above bavg threshold
+    df_stats_aoi_bavg["perc_threshold_bavg"] = (
+        df_stats_aoi_bavg_thresh[f"count_{pcode0_col}"]
+        / df_stats_aoi_bavg[f"count_{pcode0_col}"]
+        * 100
+    )
```

Comment:
Instead of computing the stats every time, can't we just count the number of non-na pixels for `da_bavg` and `da_bavg_thresh` and calculate the fraction?

Reply:
smarty! luckily our raster stats function is so versatile, that you can tell it to only count. So chose to still send it to the raster stats function, but only make it do the counting.